### PR TITLE
Handle unconnected gradients in gpflow.optimizers.Scipy.

### LINKED ISSUE #1600: Handling unconnected gradients in Scipy optimizer
# Feature request

Our Scipy optimizer wrapper should explicitly handle unconnected gradients.

## Motivation

Currently, the GPflow Scipy optimizer wrapper (`gpflow.optimizers.Scipy`) simply errors (with one of the typically unhelpful TensorFlow error messages) when one of the passed-in `variables` is not connected to the loss to be minimized. This can easily happen, for example when using an SVGP model with a `Constant` kernel (which might be used as a very simple baseline) but having inducing locations not explicitly set to non-trainable. This could also happen erroneously, e.g. when calling `Scipy().minimize(model1.training_loss, model2.trainable_variables)`.

## Proposal

**Describe the solution you would like**
An explicit error message when the variables are not connected to the loss (and hence the `grads` list contains `None` entries).

**What alternatives have you considered?**
We could also simply change the call for the gradient computation to use `unconnected_gradients=tf.UnconnectedGradients.ZERO`. This might silently hide bug cases like `minimize(model1.training_loss, model2.trainable_variables)`, so perhaps this option could be an extra flag to the minimize() call?

### PR DESCRIPTION
**Pull request by jesnie**
**PR type:** bugfix / enhancement

**Related issue(s)/PRs:** Fixes #1600

## Summary

**Proposed changes**

* Update `gpflow.optimizers.Scipy.minimize` to take a new parameter `allow_unused_variables`.
* If a variable is unused/unconnected and `allow_unused_variables=True`, then a warning is printed, but optimization continues.
* If a variable is unused/unconnected and `allow_unused_variables=False`, then an exception with an appropriate message is raised.

**What alternatives have you considered?**
This PR competes with #1655 .
I think it's an open question whether `allow_unused_variables` should default to `True` or `False`. The current suggestion is to default to `True`.

### Minimal working example

```python
import tensorflow as tf
import gpflow
from gpflow.config import default_float

target1 = [0.2, 0.8]
v1 = tf.Variable([0.5, 0.5], dtype=default_float(), name="v1")
v2 = tf.Variable([0.5], dtype=default_float(), name="v2")

def f() -> tf.Tensor:
    # v2 not used.
    return tf.reduce_sum((target1 - v1) ** 2)

opt = gpflow.optimizers.Scipy()
print(opt.minimize(f, [v1, v2], allow_unused_variables=True))
```

### Release notes

**Fully` backwards compatible:** yes

**If not, why is it worth breaking backwards compatibility:**
N/A

## PR checklist

- [x] New features: code is well-documented
  - [x] detailed docstrings (API documentation)
  - [ ] notebook examples (usage demonstration)
- [x] The bug case / new feature is covered by unit tests
- [x] Code has type annotations
- [x] Build checks
  - [x] I ran the black+isort formatter (`make format`)
  - [x] I locally tested that the tests pass (`make check-all`)
- [x] Release management
  - [x] RELEASE.md updated with entry for this change
  - [ ] New contributors: I've added myself to CONTRIBUTORS.md

### CODE REVIEW

**Review by frgsimpson — COMMENTED** [review]
Looks like a really nice improvement!

My only concern is whether it's necessary to perform these gradient checks at each call to eval. 
I suspect gradients can't become connected/disconnected during gradient descent, so we perhaps need only perform this test once for each call to 'minimise'.

File: gpflow/optimizers/scipy.py [hunk 1]
```diff
@@ -97,13 +102,32 @@ def initial_parameters(cls, variables: Sequence[tf.Variable]) -> tf.Tensor:
...
     @classmethod
     def eval_func(
-        cls, closure: LossClosure, variables: Sequence[tf.Variable], compile: bool = True
+        cls,
+        closure: LossClosure,
+        variables: Sequence[tf.Variable],
+        compile: bool = True,
+        allow_unused_variables: bool = False,
     ) -> Callable[[np.ndarray], Tuple[np.ndarray, np.ndarray]]:
+        first_call = True
```
Comment (frgsimpson) [inline gpflow/optimizers/scipy.py:111]: Does this need to be tf.constant(True), to ensure the if statement is revisited?

**Review by frgsimpson — APPROVED** [review]
LGTM! Just a small question about whether the if statement will be re-executed at each iteration.

File: gpflow/optimizers/scipy.py (hunk 1 above)
Comment (jesnie) [reply gpflow/optimizers/scipy.py:111]: > Does this need to be tf.constant(True), to ensure the if statement is revisited?

Depends... I don't really think it's necessary. If this is compiled, then we will always visit the `if first_call` branch, however all the interesting/expensive stuff inside that branch is pure Python, which won't be compiled into the TF graph, so it doesn't hurt that we're always in that branch...

Comment (frgsimpson) [reply gpflow/optimizers/scipy.py:111]: If you're sure autograph doesn't compile the enumerate call then I agree that should be fine.

Comment (jesnie) [reply gpflow/optimizers/scipy.py:111]: I wrote a small benchmark and tried using a `tf.Variable` for this if-statement - that seems to make optimisation 15-100% slower, so I'm not going to do that.
